service/dap/server.go

```diff
@@ -2490,6 +2508,35 @@ func (s *Server) resetHandlesForStoppedEvent() {
 	s.exceptionErr = nil
 }
 
+func (s *Server) runContinueCommand() (bool, *api.DebuggerState, error) {
+	s.resumeMu.Lock()
+	resumeRequestLoop := make(chan struct{})
+	defer s.asyncCommandDone(resumeRequestLoop)
+
+	go func() {
+		defer s.resumeMu.Unlock()
+		<-resumeRequestLoop
+	}()
+
+	// Check for a manual halt.
+	select {
+	case <-s.overrideStopReason:
```

Comment:
It would be more readable/reliable if this was called haltRequested or you explicitly read "manual halt" from the channel instead of relying on the comment and the fact that it's the only value ever sent over the channel.

That said, I am confused as to why this is here. Is this a manual halt that happened while next was going from one breakpoint to another? And why can't the debugger's stop reason be used?

Reply:
-> doRunCommand calls and returns from next (stopped on a breakpoint)
-> halt request is issued (debuggee is not running)
-> halt request succeeds
-> doRunCommand sees that it should continue the next request and resumes execution.

The client thinks that the program has halted, but it is actually running.

Perhaps alternatively we make sure to send a continued event so that the client knows execution was resumed? In case it thought it had halted.

That could be a simpler way to deal with this.

Reply:
I updated the description of this PR to describe the issues that we need to guard from.

This change deals with making sure that doRunCommand does not resume execution while the handleRequest goroutine expects the program to be stopped.